instrumentation-core/spring/README.md

```diff
@@ -340,6 +341,324 @@ After running Jaeger locally, navigate to the url below. Make sure to refresh th
+
+
+@RestController
+@RequestMapping(value = "/message")
+public class FirstServiceController {
+   @Autowired
+   private Tracer tracer;
+	
+   @Autowired
+   private RestTemplate restTemplate;
+
+   @Autowired
+   HttpUtils httpUtils;
+
+   private static String SECOND_SERVICE_URL = "http://localhost:8081/time";
+
+   @GetMapping
+   public String firstTracedMethod() {
+
+      ResponseEntity<String> response =
+            restTemplate.exchange(SECOND_SERVICE_URL, HttpMethod.GET, null, String.class);
+      String secondServiceTime = response.getBody();
+
+      return "Second Service says: " + secondServiceTime;
+
+   }
+}
+```
+
+As seen in the setup of SecondService, create implement the TraceInterceptor interface to wrap requests to the SecondServiceController in a span. Then register this new handler by extending the HandlerInterceptor. In effect, we will be taking a copy of the InterceptorConfig.java and ControllerTraceInterceptor.java defined in SecondService and adding it to FirstService. These files are referenced [here](#create-controller-interceptor).
+
+#### Create Client Http Request Interceptor
+
+Next, we will configure the ClientHttpRequestInterceptor to intercept all client HTTP requests made using RestTemplate.
+
+To propagate the span context from FirstService to SecondService we must inject the trace id and trace state into the outgoing request header. In section 1 this was done using the helper class HttpUtils. In this section, we will implement the ClientHttpRequestInterceptor interface and register this interceptor in our application. 
+
+Include the two classes below to your FirstService project to add this functionality:
+
+
+```java
+
+import java.io.IOException;
+
+import io.grpc.Context;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.HttpRequest;
+
+import org.springframework.http.client.ClientHttpRequestExecution;
+import org.springframework.http.client.ClientHttpRequestInterceptor;
+import org.springframework.http.client.ClientHttpResponse;
+import org.springframework.stereotype.Component;
+import io.opentelemetry.OpenTelemetry;
+import io.opentelemetry.context.propagation.HttpTextFormat;
+import io.opentelemetry.trace.Span;
+import io.opentelemetry.trace.Tracer;
+
+@Component
+public class RestTemplateHeaderModifierInterceptor implements ClientHttpRequestInterceptor {
+
+   @Autowired
+   private Tracer tracer;
+
+   private static final HttpTextFormat.Setter<HttpRequest> setter =
+         new HttpTextFormat.Setter<HttpRequest>() {
+            @Override
+            public void set(HttpRequest carrier, String key, String value) {
+               carrier.getHeaders().set(key, value);
+            }
+         };
+
+
+   @Override
+   public ClientHttpResponse intercept(HttpRequest request, byte[] body,
+         ClientHttpRequestExecution execution) throws IOException {
+
+      Span currentSpan = tracer.getCurrentSpan();
```

Comment:
this span is not ended

Reply:
I can see why this can be a bit cryptic but I started the span in ControllerTraceInterceptor.preHandle (line 442) and I end the span in ControllerTraceInterceptor.postHandle (line 454). `currentSpan` is opened and closed by the Controller Interceptor. Here I wrap each request in one span and I add events to this span in other parts of the application. I do not create sub spans with in a rest controller. In the RestTemplate interceptor defined above, I get the controller span from the tracer and I add an event. Although I'm not sure if this approach follows best practices for tracing or if I'm missing something here. Please advise and thanks again for your help.